### PR TITLE
Feature n7s tmcc#534 sequence separator

### PR DESCRIPTION
**Pull request by rphh**

### CODE REVIEW

File: tests/integration/regenerate_template.sh [hunk 1]
```diff
@@ -2,31 +2,23 @@
...
+	   then
+		   echo "CRITICAL: Cannot build project."
+		   exit 1
+	   fi
 fi
 
-echo "INFO: datamodel"
-# After build the *_datamodel.asn are saved into code subdirectory
-# This loop copies them into src subdirectory
-for datamodel in $(find -name '*_datamodel.asn')
```
Comment (Lurkerpas) [inline tests/integration/regenerate_template.sh:22]: Why is this removed?

File: tests/integration/observer_compatibility_empty/resources/taste/work/modelchecking/properties/observer/observer.pr [hunk 2]
```diff
@@ -24,23 +26,39 @@ system observer;
             errorstates error;
             successstates reached;
             /* CIF ENDTEXT */
+            /* CIF procedure (72, 421), (70, 35) */
+            /* CIF Keep Specific Geode Partition 'default' */
```
Comment (Lurkerpas) [inline tests/integration/observer_compatibility_empty/resources/taste/work/modelchecking/properties/observer/observer.pr:30]: Why this change? I though we may even drop the translation of Observable_Event....

File: tests/unit/translator/resources/variable_init.pml [hunk 3]
```diff
@@ -36,8 +36,8 @@ inline Actuator_0_transition(id)
 inline Actuator_0_init()
 {
   MyInteger_assign_value(global_state.actuator.param, 3);
-  MyInteger_assign_value(global_state.actuator.seq.x, 1);
-  MyInteger_assign_value(global_state.actuator.seq.y, 2);
+  MySeq__x_assign_value(global_state.actuator.seq.x, 1);
```
Comment (Lurkerpas) [inline tests/unit/translator/resources/variable_init.pml:39]: Does this imply that there will be many more inlines generated?

File: sdl2promela/sdl/type_info.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    def __init__(self):
+        self.all_types = {}
+
+    def fix_type_name(self, name: str) -> str:
+        if name in self.all_types:
+            return self.all_types[name]
+        return name
+
+
+# def create_type(name: str, has_parent: bool, t: Any)
```
Comment (Lurkerpas) [inline sdl2promela/sdl/type_info.py:19]: Remove?

File: sdl2promela/sdl/type_info.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+
+def build_type_info(all_types: Dict[str, Asn1Type]) -> TypeInfo:
+    type_info = TypeInfo()
+
+    for name, t in all_types.items():
+        if t.AddedType == "False":
```
Comment (Lurkerpas) [inline sdl2promela/sdl/type_info.py:26]: I was confused, because all_types is declared as Dict[str,str].... and then here as Dict[str, Asn1Type]

File: tests/integration/regenerate_template.sh (hunk 1 above)
Comment (rphh) [reply tests/integration/regenerate_template.sh:22]: It is not necessary to copy files from code to src. Now Opengeode does this.

File: sdl2promela/sdl/type_info.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        return name
+
+
+# def create_type(name: str, has_parent: bool, t: Any)
+
+
+def build_type_info(all_types: Dict[str, Asn1Type]) -> TypeInfo:
+    type_info = TypeInfo()
+
+    for name, t in all_types.items():
```
Comment (Lurkerpas) [inline sdl2promela/sdl/type_info.py:25]: Use type_name instead of name? t does not mean much.

Comment (Lurkerpas) [reply sdl2promela/sdl/type_info.py:25]: Sorry, asn1_type?

File: sdl2promela/sdl/type_info.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        if t.AddedType == "False":
+            type_info.all_types[name] = name.replace("-", "_")
+
+    for name, t in all_types.items():
+        escaped_name = name.replace("-", "_")
+        if t.AddedType == "True":
+            for b in iter(type_info.all_types):
```
Comment (Lurkerpas) [inline sdl2promela/sdl/type_info.py:32]: Use something more descriptive than b.

File: sdl2promela/sdl/type_info.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+                if escaped_name.startswith(b):
+                    part_name = escaped_name.removeprefix(
+                        b
+                    )  # part_name shall start with '_'
+                    type_info.all_types[escaped_name] = f"{b}_{part_name}"
```
Comment (Lurkerpas) [inline sdl2promela/sdl/type_info.py:37]: OK, so this basically replaces prefix_name with prefix__name, right? Can we have a comment about the intent of the method.... or maybe do an extraction to avoid a comment?

File: sdl2promela/sdl/type_info.py (hunk 5 above)
Comment (rphh) [reply sdl2promela/sdl/type_info.py:26]: the parameter `all_types` is `Dict[str, Asn1Type]`
The `TypeInfo` class has `all_types` which is `Dict[str, str]`

File: tests/unit/translator/resources/variable_init.pml (hunk 3 above)
Comment (rphh) [reply tests/unit/translator/resources/variable_init.pml:39]: No, the number of generated inlines is not changed. The generator uses type from left operand to create name of the inline.

File: tests/integration/observer_compatibility_empty/resources/taste/work/modelchecking/properties/observer/observer.pr (hunk 2 above)
Comment (rphh) [reply tests/integration/observer_compatibility_empty/resources/taste/work/modelchecking/properties/observer/observer.pr:30]: Well, we had a discussion about it and the conclusion was that the the procedure is not required to translate observer with sdl2promela.
However, after some time I realised that I added `observe` procedures for tests with `_compatibility_` in name.
Those tests check if SDL created using space creator can be translated using sdl2promela.
At this moment I have no idea why those procedures are missing, there is only a declaration, but no body.

I suggest that empty procedure `observe` is required here.

File: sdl2promela/sdl/type_info.py (hunk 4 above)
Comment (rphh) [reply sdl2promela/sdl/type_info.py]: Done

File: sdl2promela/sdl/type_info.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+
+def build_type_info(all_types: Dict[str, Asn1Type]) -> TypeInfo:
+    type_info = TypeInfo()
+
+    for name, t in all_types.items():
+        if t.AddedType == "False":
+            type_info.all_types[name] = name.replace("-", "_")
+
+    for name, t in all_types.items():
```
[the remark this answers is not in the document]
Comment (rphh) [reply sdl2promela/sdl/type_info.py]: I added documentation and a lot of comments.

File: sdl2promela/sdl/type_info.py (hunk 6 above)
Comment (rphh) [reply sdl2promela/sdl/type_info.py]: Changed

File: sdl2promela/sdl/type_info.py (hunk 7 above)
Comment (rphh) [reply sdl2promela/sdl/type_info.py]: Changed

File: sdl2promela/sdl/type_info.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (rphh) [reply sdl2promela/sdl/type_info.py]: In fact, I had to refactor this to support nesting with more than one level.

**Review by Lurkerpas — APPROVED** [review]

**Review by Lurkerpas — APPROVED** [review]
